thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/MeanVarianceRuleDetector.java

```diff
@@ -0,0 +1,277 @@
+    double[] errorArray = new double[size];
+    double[] std = new double[size];
+    double[] mean = new double[size];
+
+    //get the trainingDF for each week, which is the number of lookback to 1 week before the each week predict start time
+    for (int k = 0; k < size; k++) {
+      DataFrame trainingDF;
+      trainingDF = getLookbackDF(inputDF, forecastDF.getLong(COL_TIME, k));
+      //the get historical WoW mean and std.
+      std[k]= trainingDF.getDoubles(COL_CHANGE).std().value();
+      mean[k] = trainingDF.getDoubles(COL_CHANGE).mean().value();
+
+      //calculate baseline, error , upper and lower bound for prediction window.
+      resultTimeArray[k] = forecastDF.getLong(COL_TIME, k);
+      baselineArray[k] = trainingDF.getDouble(COL_VALUE,trainingDF.size()-1) * (1 + mean[k]);
+      errorArray[k] = trainingDF.getDouble(COL_VALUE,trainingDF.size()-1) * sigma * std[k];
+      upperBoundArray[k] = baselineArray[k] + errorArray[k];
+      lowerBoundArray[k] = baselineArray[k] - errorArray[k];
+    }
+    //Construct the dataframe.
+    resultDF.addSeries(COL_TIME, LongSeries.buildFrom(resultTimeArray)).setIndex(COL_TIME);
+    resultDF.addSeries(COL_VALUE, DoubleSeries.buildFrom(baselineArray));
+    resultDF.addSeries(COL_UPPER_BOUND, DoubleSeries.buildFrom(upperBoundArray));
+    resultDF.addSeries(COL_LOWER_BOUND, DoubleSeries.buildFrom(lowerBoundArray));
+    resultDF.addSeries(COL_ERROR, DoubleSeries.buildFrom(errorArray));
+    return resultDF;
+  }
+
+  /**
+   * Fetch data from metric
+   *
+   * @param metricEntity metric entity
+   * @param start start timestamp
+   * @param end end timestamp
+   * @return Data Frame that has data from start to end
+   */
+  private DataFrame fetchData(MetricEntity metricEntity, long start, long end) {
+    List<MetricSlice> slices = new ArrayList<>();
+    MetricSlice sliceData = MetricSlice.from(metricEntity.getId(), start, end,
+        metricEntity.getFilters(), timeGranularity);
+    slices.add(sliceData);
+    LOG.info("Getting data for" + sliceData.toString());
+    InputData data = this.dataFetcher.fetchData(new InputDataSpec().withTimeseriesSlices(slices));
+    return data.getTimeseries().get(sliceData);
+  }
+
+  /**
+   * Returns a data frame containing lookback number of data before prediction time
+   * @param originalDF the original dataframe
+   * @param time the prediction time, in unix timestamp
+   * @return DataFrame containing lookback number of data
+   */
+  private DataFrame getLookbackDF(DataFrame originalDF, Long time) {
+    LongSeries longSeries = (LongSeries) originalDF.get(COL_TIME);
+    int indexEnd = longSeries.find(time);
+    DataFrame df = DataFrame.builder(COL_TIME, COL_VALUE).build();
+
+    if (indexEnd != -1) {
+      int indexStart = Math.max(0, indexEnd - lookback);
+      df = df.append(originalDF.slice(indexStart, indexEnd));
+    }
+    // calculate percentage change
+    df.addSeries(COL_CURR,df.getDoubles(COL_VALUE).shift(-1));
```

Comment:
Just curious, why you want to calculate the percentage change, and then use the mean/std from the percentage change, instead of directly using the values?
Directly use the values can give you one more data point.

Also have you considered removing outliers? If there is one anomaly happens and this can affect the upper/lower boundaries for multiple weeks after it.

Reply:
For our use, we typically track metrics' WoW change and report that. It's what we actually tracking.

After discussion for the outlier removal, we decided to put it into phase 2. Phase one is just to have a detection method using the standard deviation that works for weekly.